Add account threat detection and storage malware scanning features (#1013)
navapbc/template-infra: docs/infra/threat-detection.md

- **DNS logs** - Domain name resolution requests

## Malware detection for S3 storage

The infratructure leverages GuardDuty's malware detection feature to continuously scans files uploaded to S3 buckets for malicious content. When malware is detected:

1. **File access is blocked** - Downloads of infected files are prevented
2. **Findings are generated** - Security findings are created in the GuardDuty service with detailed information including:
   - **Finding ID** - Unique identifier for the security event
   - **Severity level** - Low, Medium, High, Critical
   - **Finding type** - Specific threat classification (e.g., `Malware:S3/MaliciousFile`)
   - **Resource details** - Affected S3 bucket, object key, and account information
   - **Timestamp** - When the malware was detected
   - **Evidence** - Technical details about the malicious content
   - **Remediation** - Recommended actions to address the threat
3. **Tags are applied** - S3 objects are tagged with scan results for tracking:
   - **`GuardDutyMalwareScanStatus`** - Scan result status (`NO_THREATS_FOUND`, `THREATS_FOUND`)

### Checking S3 Object Tags for Malware Status

```bash

> #!/bin/bash

Reviewer: ```suggestion
#!/usr/bin/env bash
```

Author: Done

# Threat Detection (AWS GuardDuty)

> The infrastructure uses AWS GuardDuty for threat detection and security monitoring across the AWS account. GuardDuty is enabled by default to continuously analyze the following data sources:

Reviewer: ```suggestion
The infrastructure uses AWS GuardDuty for threat detection and security monitoring within AWS accounts. GuardDuty is enabled by default to continuously analyze the following data sources:
```

Is more accurate I think?

Author: Done

    --query "MalwareProtectionPlans[*].MalwareProtectionPlanId" \
    --output text); do

    bucket=$(aws guardduty get-malware-protection-plan \
        --malware-protection-plan-id "$plan_id" \
        --query "ProtectedResource.S3Bucket.BucketName" \
        --output text 2>/dev/null)

    if [ "$bucket" != "None" ] && [ -n "$bucket" ]; then
        echo "Scanning protected bucket: $bucket ..."

        aws s3api list-objects-v2 --bucket "$bucket" \
            --query "Contents[*].Key" \
            --output text 2>/dev/null | tr '\t' '\n' | \
        while read -r key; do
            if [ -n "$key" ]; then
                tags=$(aws s3api get-object-tagging \
                    --bucket "$bucket" \
                    --key "$key" \
                    --query "TagSet[?Key=='GuardDutyMalwareScanStatus' && Value=='THREATS_FOUND'].Value" \
                    --output text 2>/dev/null)

                if [ "$tags" = "THREATS_FOUND" ]; then
                    echo "MALWARE DETECTED: s3://$bucket/$key"
                fi
            fi
        done
    fi
done
```

### Accessing GuardDuty Findings

**AWS Console:**

> - Navigate to GuardDuty Console → Findings

Reviewer: ```suggestion
- Navigate to GuardDuty Console -> Findings
```

Author: Changed

- Filter by finding type, severity, or resource
- View detailed finding information and evidence

## Configuration

> **Threat detection is enabled by default** for all environments. The configuration can be customized through Terraform variables in the accounts layer.

Reviewer: We should clarify it's only supported for the `project_config.default_region`, so if they need to support multiple regions do X (which I guess would be adding new providers for their desired regions and `source ../modules/threat_detection` blocks for each new provider?).

Author: Changed.

To disable AWS GuardDuty threat detection:

1. Edit your Terraform workspace configuration file for the account layer, `infra/project-config/threat_detection.tf`

2. Set the threat detection variable to `false`:
   ```hcl
   enable_threat_detection = false
   ```

3. Set finding publishing frequency (FIFTEEN_MINUTES, ONE_HOUR, SIX_HOURS)
   ```hcl
   threat_detection_finding_publishing_frequency = "FIFTEEN_MINUTES"
   ```
4. Apply the changes:
   ```bash
   make infra-update-current-account
   ```

### Available Configuration Options

| Variable | Description | Default | Options |
|----------|-------------|---------|---------|
| `enable_threat_detection` | Enable/disable GuardDuty threat detection | `true` | `true`, `false` |
| `threat_detection_finding_publishing_frequency` | How often GuardDuty publishes findings to CloudWatch Events | `"FIFTEEN_MINUTES"` | `"FIFTEEN_MINUTES"`, `"ONE_HOUR"`, `"SIX_HOURS"` |

### Enabling malware detection for S3 storage

1. Edit your Terraform configuration file for the application service layer, `infra/<application-name>/app-config/main.tf`

2. Set the `enable_storage_malware_scanning` variable to `true`

3. Apply the changes:
   ```bash

>    make infra-update-app-service APP_NAME=<application-name> ENVIRONMENT=<ENVIRONMENT>

Reviewer: ```suggestion
   make infra-update-app-service APP_NAME=<APP_NAME> ENVIRONMENT=<ENVIRONMENT>
```

2. Set the threat detection variable to `false`:

   ```hcl
   enable_threat_detection = false
   ```

3. Set finding publishing frequency (FIFTEEN_MINUTES, ONE_HOUR, SIX_HOURS)
   ```hcl
   threat_detection_finding_publishing_frequency = "FIFTEEN_MINUTES"
   ```
4. Apply the changes:
   ```bash
   make infra-update-current-account
   ```

### Available Configuration Options

| Variable                                        | Description                                                 | Default             | Options                                          |
| ----------------------------------------------- | ----------------------------------------------------------- | ------------------- | ------------------------------------------------ |
| `enable_threat_detection`                       | Enable/disable GuardDuty threat detection                   | `true`              | `true`, `false`                                  |
| `threat_detection_finding_publishing_frequency` | How often GuardDuty publishes findings to CloudWatch Events | `"FIFTEEN_MINUTES"` | `"FIFTEEN_MINUTES"`, `"ONE_HOUR"`, `"SIX_HOURS"` |

### Enabling malware detection for S3 storage

> 1. Edit your Terraform configuration file for the application service layer, `infra/<application-name>/app-config/main.tf`

Reviewer: ```suggestion
1. Edit your Terraform configuration file for the application service layer, `infra/<APP_NAME>/app-config/main.tf`
```

Author: Changed.
